# Q1

# Task:
    String Normalizer & Finder. 
    Write a function normalize_and_find(text: str, needle: str) -> int that: (1) trims whitespace on both ends, (2) converts to lowercase, then (3) returns the index of the first occurrence of needle using a safe search (return −1 if not found). Discuss alternatives (find vs. index), and how your choice affects error handling. Anchor concepts: stripping, case conversion, find/replace, string immutability. Reference: :contentReference[oaicite:6]index=6

# Prompt:
    Write a Python function normalize_and_find(text, needle) that trims spaces, ignores case, and returns the index of the search word. Use find() so it returns -1 if not found. Add a quick test.

In [4]:
#ChatGPT Code

def normalize_and_find(text, needle):
    text = text.strip().lower()
    needle = needle.lower()
    return text.find(needle)

print(normalize_and_find("   Hello, World!  ", "world"))

7


# Critique:
    Correctness: The function correctly normalizes text and searches for needle, returning the correct index or -1 if not found.
    Complexity: Time complexity is O(n + m), where n is the length of text and m is the length of needle. Space complexity is O(1) since no extra large data structures are created.
    Robustness: The function handles typical cases but could be improved with input validation to handle None values for text or needle.
    Readability: The code is simple and easy to follow, but a docstring explaining its purpose and adding more test cases would improve clarity.
    Faithfulness: The solution uses .strip() and .lower() for normalization and find() for safe searching, as required by the prompt. The behavior matches the spec perfectly.

In [5]:
#Improved Code

def normalize_and_find(text, needle):
    return text.strip().lower().find(needle.lower())

print(normalize_and_find("   Hello, World!  ", "world"))

7


# Explaination:
    find() is safer because it returns -1 if the word isn't found, without raising an error. index() raises an error if the word is missing, so you need extra code to handle that.

# Q2

# Task:
    Leetspeak Translator (Parametric). Generalize our leetspeak idea: write translate(text: str, mapping: dict[str,str]) -> str that applies a mapping (e.g., {’a’:’4’,’e’:’3’,...}) over lowercase letters only, leaving others unchanged. Explain why strings are immutable and show an efficient approach (e.g., list join vs. repeated concatenation). Anchor concepts: replace, iteration strategies, efficiency. Reference: :contentReference[oaicite:7]index=7


# Prompt:
    Write a Python function translate(text, mapping) that replaces lowercase letters in text using the given mapping dictionary. Characters not in the mapping should stay the same. Return the translated string. Use an efficient method and add a quick test.

In [6]:
#ChatGPT Code

def translate(text, mapping):
    return "".join(mapping.get(char, char) for char in text)

print(translate("apple", {'a': '4', 'e': '3'}))

4ppl3


# Critique
    Correctness: The function correctly applies the mapping dictionary to replace lowercase letters, leaving other characters unchanged, as specified by the prompt.
    Complexity: Time complexity is O(n), where n is the length of text, because the function iterates through the string once. Space complexity is O(n) due to the creation of a new string using "".join().
    Robustness: The function works for typical inputs but doesn't handle cases where text or mapping could be None. Adding basic input validation would improve robustness.
    Readability: The code is concise and easy to read. Using .join() with a generator expression is efficient and clear, making it a good choice for transforming the string.
    Faithfulness: The solution adheres to the prompt by using an efficient string-building method ("".join()), and the mapping is applied correctly to lowercase letters. The behavior is consistent with the specification.

In [7]:
#Improved Code

def translate(text, mapping):
    result = ""
    for ch in text:
        if ch in mapping:
            result += mapping[ch]
        else:
            result += ch
    return result

print(translate("apple", {'a': '4', 'e': '3'}))

4ppl3


# Explaination:
    Strings in Python can’t be changed once created, which helps with safety and memory. Concatenating strings with += is slow, but using a list and then joining is faster.

# Q3

# Task: 
    Currency Formatter & Rounding Ties-to-Even. Write fmt_money(x: float) -> str that returns a string like 1,234.50 using proper rounding (banker’s rounding / ties-to-even). Include test cases that demonstrate ties (e.g., 2.5, 3.5). Explain 2 floating-point representation error and how formatting mitigates it for display. Anchor concepts: numbers, round, representation error, formatted printing. Reference: :contentReference[oaicite:8]index=8

# Prompt:
    Write a Python function fmt_money(x) that returns a float formatted with commas and 2 decimal places using banker’s rounding (round ties to even). Add test cases for 2.5 and 3.5. Explain how formatting helps with floating-point errors.

In [8]:
#ChatGPT Code

def fmt_money(x):
    rounded = round(x + 1e-8, 2)
    whole, dot, decimal = f"{rounded:.2f}".partition(".")
    reversed_whole = whole[::-1]
    grouped = [reversed_whole[i:i+3] for i in range(0, len(reversed_whole), 3)]
    with_commas = ",".join(grouped)[::-1]
    return with_commas + "." + decimal

print(fmt_money(1234.5))
print(fmt_money(2.5))
print(fmt_money(3.5))

1,234.50
2.50
3.50


# Critique
    Correctness: The function correctly formats numbers with commas and two decimal places, using banker’s rounding. The rounding trick (+ 1e-8) is unnecessary and could cause issues with extreme values.
    Complexity: Time complexity is O(n) for string operations. Space complexity is O(n) due to the string manipulations.
    Robustness: Handles typical cases well, but no input validation (e.g., for non-numeric inputs) is present.
    Readability: Code is functional but can be simplified. The string reversal and grouping for commas are a bit complex.
    Faithfulness: Correctly implements banker’s rounding and addresses floating-point representation issues, though a simpler approach could improve clarity.

In [9]:
#Improved Code

def fmt_money(x):
    return f"{x:,.2f}"

print(fmt_money(1234.5)) 
print(fmt_money(2.5))    
print(fmt_money(3.5))     

1,234.50
2.50
3.50


# Explaination:
    Floating-point errors happen because computers can’t store decimal numbers perfectly. Formatting like f"{x:,.2f}" fixes this by rounding the number to 2 decimal places, so even if there's a tiny error, it shows cleanly (e.g., 2.50 instead of 2.499999).

# Q4

# Task:
    Exponent Tool & Input Validation. Create a CLI program exponent.py that reads two values, validates they are numeric, handles errors gracefully (e.g., ValueError), and prints a ** b with clear messages. Compare using try/except vs. pre-validation with str.isdigit() (and why isdigit() is insufficient for negatives/floats). Anchor concepts: arithmetic operators, exceptions, input. Reference: :contentReference[oaicite:9]index=9, :contentReference[oaicite:10]index=10

# Prompt:
    Write a CLI program that reads two numbers from input, validates them, and prints a ** b. Use error handling to catch invalid input. Compare try/except with using isdigit() and explain why isdigit() can’t handle negatives or floats well.

In [10]:
#ChatGPT Code

def get_num(prompt, retries=3):
    for _ in range(retries):
        val = input(prompt).strip()
        if val.lower() == 'q':
            exit("Quit.")
        try:
            return float(val)
        except:
            print("Invalid input.")
    exit("Too many tries.")

a = get_num("Base: ")
b = get_num("Exponent: ")
try:
    print(f"{a} ** {b} = {a ** b}")
except Exception as e:
    print(f"Error: {e}")

Base:  3
Exponent:  4


3.0 ** 4.0 = 81.0


# Critique
    Correctness: Works as expected, but the except block is too general. Catch specific exceptions (like ValueError) for clarity.
    Complexity: The program is efficient with O(1) time and space complexity for both input handling and exponentiation.
    Robustness: Handles invalid input well, but using str.isdigit() could be more limiting for negative numbers and floats. Better to validate input more precisely.
    Readability: The code is clear, but could be improved with more specific exception handling and clearer error messages.
    Faithfulness: Uses try/except correctly, but comparing it with str.isdigit() would help demonstrate the limitations of isdigit() for non-integer values.

In [11]:
#Improved Code

def main():
    base = input("Enter the base number: ")
    exponent = input("Enter the exponent number: ")

    try:
        base = float(base)
        exponent = float(exponent) 
        result = base ** exponent
        print(f"{base} raised to the power of {exponent} is {result}")
    except ValueError:
        print("Oops! Please enter valid numbers.")

if __name__ == "__main__":
    main()

Enter the base number:  3
Enter the exponent number:  4


3.0 raised to the power of 4.0 is 81.0


# Explaination:
    Using str.isdigit() just checks if the input is a number made up of digits, so it works fine for positive whole numbers but fails with negative numbers or decimals. On the other hand, try/except tries to convert the input into a number and catches any errors if it’s not valid, so it works for any number, even with decimals or negatives, but it’s a bit more complicated.

# Q5

# Task:
    Password Policy Checker (If/Elif/Else + Loops). Implement check_password(pw: str) -> dict that validates: length ≥ 10, at least one uppercase, one lowercase, one digit, one symbol in !@#$%^&*, and no spaces. Return a dict of booleans per rule and an overall pass/fail. Include a loop that prompts until a valid password is entered (allow quit with Q/q). Anchor concepts: conditionals, logical operators, loops, break/continue. Reference: :contentReference[oaicite:11]index=11

# Prompt: 
    Write a function check_password(pw) that checks a password against multiple rules: length, uppercase, lowercase, digit, symbol, no spaces. Return a dict of booleans for each rule and overall pass/fail. Keep prompting user until password is valid or user quits.

In [12]:
#ChatGPT Code

def check_password(pw):
    result = {
        "length": len(pw) >= 10,
        "uppercase": any(c.isupper() for c in pw),
        "lowercase": any(c.islower() for c in pw),
        "digit": any(c.isdigit() for c in pw),
        "symbol": any(c in "!@#$%^&*" for c in pw),
        "no_spaces": " " not in pw,
    }
    result["passed"] = all(result.values())
    return result

while True:
    pw = input("Enter password (Q to quit): ")
    if pw.lower() == "q":
        break
    res = check_password(pw)
    if res["passed"]:
        print("Password valid!")
        break
    else:
        print("Password invalid, check rules:", res)

Enter password (Q to quit):  PasswordVali123!


Password valid!


# Critique
    Correctness: The function correctly checks the password against the specified rules and returns a dictionary of booleans for each rule, including an overall pass/fail status. The loop effectively prompts the user until a valid password is entered or the user chooses to quit.
    Complexity: The time complexity is O(n) for each password check, where n is the length of the password. The space complexity is O(1) since only a fixed number of checks are performed regardless of input size.
    Robustness: The function works for typical cases, but it could be improved by handling edge cases like empty passwords. Also, the program does not enforce any limits on the number of attempts before quitting.
    Readability: The code is straightforward, but it could benefit from clearer variable names (e.g., pw to password) and comments explaining the checks. The logic is simple and easy to follow.
    Faithfulness: The function uses conditionals and logical operators to check the password according to the prompt. The loop correctly prompts until a valid password is entered or the user quits, meeting the requirements of the task.

In [13]:
# Improved Code

def check_password(pw):
    if len(pw) < 10:
        return False
    if " " in pw:
        return False
    if not any(c.isupper() for c in pw):
        return False
    if not any(c.islower() for c in pw):
        return False
    if not any(c.isdigit() for c in pw):
        return False
    if not any(c in "!@#$%^&*" for c in pw):
        return False
    return True

while True:
    pw = input("Enter password (Q to quit): ")
    if pw.lower() == "q":
        break
    if check_password(pw):
        print("Password valid!")
        break
    else:
        print("Password invalid. Make sure it’s 10+ chars, has upper & lower case letters, digits, symbols, and no spaces.")

Enter password (Q to quit):  PasswordValid123!


Password valid!


# Q6

# Task:
    Dice Simulation (Monte Carlo). Write roll() -> int to simulate a fair die, then simulate 100,000 rolls to estimate the probability that the sum of two dice equals 7. Compare empirical result to the exact probability. Comment on random seeding and reproducibility. Anchor concepts: random module, simulation, averages. Reference: :contentReference[oaicite:12]index=12

# Prompt:
    Write a function roll() to stimulate fair die roll. Then simulate 100,000 rolls and calculate the probability that the sum of two dice equals 7. Compare the result with the exact probability.

In [14]:
# ChatGPT Code

import random

def simulate_rolls(num_rolls=100000):
    count_seven = sum(1 for _ in range(num_rolls) if random.randint(1, 6) + random.randint(1, 6) == 7)
    return count_seven / num_rolls

if __name__ == "__main__":
    random.seed(42)
    empirical_prob = simulate_rolls()
    print(f"Empirical Probability of sum = 7: {empirical_prob:.4f}")
    print(f"Exact Probability of sum = 7: {1/6:.4f}")

Empirical Probability of sum = 7: 0.1646
Exact Probability of sum = 7: 0.1667


# Critique
    Correctness: Correctly simulates dice rolls and calculates the empirical probability of the sum being 7. Compares it with the exact probability.
    Complexity: O(n) time complexity for num_rolls rolls, O(1) space.
    Robustness: Works well, but the function doesn't handle invalid inputs for num_rolls (e.g., non-integer or negative values).
    Readability: Clear code. The use of sum() with a generator expression is efficient and easy to understand.
    Faithfulness: Meets the prompt requirements, simulating the rolls and comparing results with the exact probability.

In [15]:
# Improved Code

import random

def roll():
    return random.randint(1, 6)

def simulate_rolls(num_rolls=100000):
    count_seven = 0
    for _ in range(num_rolls):
        if roll() + roll() == 7:
            count_seven += 1
    return count_seven / num_rolls

if __name__ == "__main__":
    random.seed(42)
    empirical_prob = simulate_rolls()
    print(f"Empirical Probability of sum = 7: {empirical_prob:.4f}")
    print(f"Exact Probability of sum = 7: {6/36:.4f}")

Empirical Probability of sum = 7: 0.1646
Exact Probability of sum = 7: 0.1667


# Q7

# Task
    OOP: Vehicles with Odometer and Fuel. Design classes Vehicle (base), Car and Truck (children). Each has color, mileage, and fuel_liters. Provide drive(km, km_per_liter) which increments mileage and decreases fuel (never below 0) and refuel(liters). Show how inheritance avoids duplication and add a method override in Truck (e.g., different efficiency). Include a minimal test zscript. Anchor concepts: classes vs. instances, attributes, instance methods, inheritance/override. Reference: :contentReference[oaicite:13]index=13

# Prompt:
    Create a vehicle class with attributes color, mileage, and fuel_liters. Implement drive(km, km_per_liter) to update mileage and decreased fuel, and refuel(liters) to add fuel. Derive Car and Truck classes with different fuel efficiences. Override the drive() method in Truck to reflect its efficiency. Include a test script to check the functionality|

In [16]:
#ChatGPT Code

class Vehicle:
    def __init__(self, color, mileage=0, fuel_liters=0):
        self.color = color
        self.mileage = mileage
        self.fuel_liters = fuel_liters

    def drive(self, km, km_per_liter):
        fuel_needed = km / km_per_liter
        if self.fuel_liters >= fuel_needed:
            self.fuel_liters -= fuel_needed
            self.mileage += km
        else:
            print(f"Not enough fuel to drive {km} km. Please refuel.")

    def refuel(self, liters):
        self.fuel_liters += liters
        print(f"Refueled {liters} liters. Current fuel: {self.fuel_liters} liters.")

    def get_status(self):
        return f"Color: {self.color}, Mileage: {self.mileage} km, Fuel: {self.fuel_liters} liters"


class Car(Vehicle):
    def __init__(self, color, mileage=0, fuel_liters=0, efficiency=15):
        super().__init__(color, mileage, fuel_liters)
        self.efficiency = efficiency

    def drive(self, km):
        super().drive(km, self.efficiency)


class Truck(Vehicle):
    def __init__(self, color, mileage=0, fuel_liters=0, efficiency=10):
        super().__init__(color, mileage, fuel_liters)
        self.efficiency = efficiency

    def drive(self, km):
        super().drive(km, self.efficiency)


if __name__ == "__main__":
    car = Car(color="Red", fuel_liters=50)
    truck = Truck(color="Blue", fuel_liters=100)

    print("Initial Status:")
    print(car.get_status())
    print(truck.get_status())

    print("\nDriving the Car for 100 km...")
    car.drive(100)
    print(car.get_status())

    print("\nDriving the Truck for 100 km...")
    truck.drive(100)
    print(truck.get_status())

    print("\nRefueling the Car by 20 liters...")
    car.refuel(20)
    print(car.get_status())

    print("\nRefueling the Truck by 50 liters...")
    truck.refuel(50)
    print(truck.get_status())

Initial Status:
Color: Red, Mileage: 0 km, Fuel: 50 liters
Color: Blue, Mileage: 0 km, Fuel: 100 liters

Driving the Car for 100 km...
Color: Red, Mileage: 100 km, Fuel: 43.333333333333336 liters

Driving the Truck for 100 km...
Color: Blue, Mileage: 100 km, Fuel: 90.0 liters

Refueling the Car by 20 liters...
Refueled 20 liters. Current fuel: 63.333333333333336 liters.
Color: Red, Mileage: 100 km, Fuel: 63.333333333333336 liters

Refueling the Truck by 50 liters...
Refueled 50 liters. Current fuel: 140.0 liters.
Color: Blue, Mileage: 100 km, Fuel: 140.0 liters


# Critique
    Correctness: The code correctly models the behavior of vehicles (Car and Truck), including driving and refueling with proper fuel consumption and mileage updates. The drive method works as expected with the proper fuel consumption based on the efficiency.
    Complexity: The program is simple, with O(1) time and space complexity for each method call (drive, refuel, get_status).
    Robustness: Works as expected for basic use cases. It doesn't handle edge cases such as invalid input for fuel amount or kilometers (e.g., negative values or non-numeric inputs).
    Readability: The code is clear and easy to understand. Using super() to call the parent class methods is efficient and avoids code repetition.
    Faithfulness: The code is faithful to the task, using class inheritance and method overriding for Car and Truck classes. It encapsulates vehicle behavior (fuel consumption, mileage, refueling) appropriately.

In [17]:
# Improved Code

class Vehicle:
    def __init__(self, color, fuel_liters=0, mileage=0, efficiency=0):
        self.color = color
        self.fuel_liters = fuel_liters
        self.mileage = mileage
        self.efficiency = efficiency

    def drive(self, km):
        fuel_needed = km / self.efficiency
        if self.fuel_liters >= fuel_needed:
            self.fuel_liters -= fuel_needed
            self.mileage += km
        else:
            print("Not enough fuel to drive. Please refuel.")

    def refuel(self, liters):
        self.fuel_liters += liters
        print(f"Refueled {liters} liters.")

    def get_status(self):
        return f"{self.color} vehicle, Mileage: {self.mileage} km, Fuel: {self.fuel_liters} liters"

class Car(Vehicle):
    def __init__(self, color, fuel_liters=0, mileage=0):
        super().__init__(color, fuel_liters, mileage, efficiency=15)

class Truck(Vehicle):
    def __init__(self, color, fuel_liters=0, mileage=0):
        super().__init__(color, fuel_liters, mileage, efficiency=10)

if __name__ == "__main__":
    car = Car(color="Red", fuel_liters=50)
    truck = Truck(color="Blue", fuel_liters=100)

    print(car.get_status())
    print(truck.get_status())

    car.drive(100)
    print(car.get_status())

    truck.drive(100)
    print(truck.get_status())

    car.refuel(20)
    print(car.get_status())

    truck.refuel(50)
    print(truck.get_status())


Red vehicle, Mileage: 0 km, Fuel: 50 liters
Blue vehicle, Mileage: 0 km, Fuel: 100 liters
Red vehicle, Mileage: 100 km, Fuel: 43.333333333333336 liters
Blue vehicle, Mileage: 100 km, Fuel: 90.0 liters
Refueled 20 liters.
Red vehicle, Mileage: 100 km, Fuel: 63.333333333333336 liters
Refueled 50 liters.
Blue vehicle, Mileage: 100 km, Fuel: 140.0 liters


# Q8

# Task
    OOP + Composition: Kennel with Dogs. Using a Dog class (name, age, breed, bark()), implement a Kennel manager that holds multiple Dog instances, supports add_dog, remove_dog, oldest_dog(), and find_by_breed(breed). Compare composition vs. inheritance for this scenario. Anchor concepts: OOP design, managing collections of objects. Reference: :contentReference[oaicite:14]index=14

# Prompt
    Create a dog class with attributes name, age, breed, and a method bark(). Implement a Kennel class that manages multiple Dog instances. The Kennel class should have methods to:
    add_dog(dog)
    remove_dog(dog)
    oldest_dog()
    find_by_breed(breed)

In [18]:
#ChatGPT Code

class Dog:
    def __init__(self, name, age, breed):
        self.name = name
        self.age = age
        self.breed = breed

    def bark(self):
        return f"{self.name} says woof!"

class Kennel:
    def __init__(self):
        self.dogs = []

    def add_dog(self, dog):
        self.dogs.append(dog)

    def remove_dog(self, dog):
        if dog in self.dogs:
            self.dogs.remove(dog)

    def oldest_dog(self):
        if self.dogs:
            return max(self.dogs, key=lambda dog: dog.age)
        return None

    def find_by_breed(self, breed):
        return [dog for dog in self.dogs if dog.breed == breed]

# Test the code
if __name__ == "__main__":
    dog1 = Dog("Buddy", 5, "Golden Retriever")
    dog2 = Dog("Bella", 3, "Labrador")
    dog3 = Dog("Max", 8, "Beagle")
    
    kennel = Kennel()
    
    kennel.add_dog(dog1)
    kennel.add_dog(dog2)
    kennel.add_dog(dog3)
    
    print("All dogs in the kennel:")
    for dog in kennel.dogs:
        print(f"{dog.name} ({dog.age} years old)")

    print("\nOldest dog:")
    oldest = kennel.oldest_dog()
    if oldest:
        print(f"{oldest.name} ({oldest.age} years old)")
    
    print("\nDogs of breed 'Labrador':")
    labrador_dogs = kennel.find_by_breed("Labrador")
    for dog in labrador_dogs:
        print(dog.name)

    print("\nRemoving 'Max' from the kennel...")
    kennel.remove_dog(dog3)

    print("\nAll dogs in the kennel after removal:")
    for dog in kennel.dogs:
        print(f"{dog.name} ({dog.age} years old)")

All dogs in the kennel:
Buddy (5 years old)
Bella (3 years old)
Max (8 years old)

Oldest dog:
Max (8 years old)

Dogs of breed 'Labrador':
Bella

Removing 'Max' from the kennel...

All dogs in the kennel after removal:
Buddy (5 years old)
Bella (3 years old)


# Critique
    Correctness: The code works well for managing dogs in a kennel, with correct methods for adding, removing, and finding dogs.
    Complexity: Time complexity is O(n) for oldest_dog() and find_by_breed(), which is fine for the operations being performed.
    Robustness: It lacks input validation for attributes like age. If a dog isn’t in the kennel, the remove_dog() method silently ignores it.
    Readability: Code is clean and easy to understand. Adding docstrings would improve clarity.
    Faithfulness: The code matches the intended functionality, accurately modeling the kennel operations.

In [19]:
#Improved Code

class Dog:
    def __init__(self, name, age, breed):
        self.name = name
        self.age = age
        self.breed = breed

    def bark(self):
        return f"{self.name} says woof!"
    
    def __str__(self):
        return f"{self.name} ({self.age} years old, {self.breed})"

class Kennel:
    def __init__(self):
        self.dogs = []

    def add_dog(self, dog):
        self.dogs.append(dog)

    def remove_dog(self, dog):
        if dog in self.dogs:
            self.dogs.remove(dog)
        else:
            print(f"{dog.name} is not in the kennel!")

    def oldest_dog(self):
        if not self.dogs:
            print("No dogs in the kennel.")
            return None
        return max(self.dogs, key=lambda dog: dog.age)

    def find_by_breed(self, breed):
        return [dog for dog in self.dogs if dog.breed == breed]

    def __str__(self):
        return f"Kennel has {len(self.dogs)} dogs."

if __name__ == "__main__":
    dog1 = Dog("Buddy", 5, "Golden Retriever")
    dog2 = Dog("Bella", 3, "Labrador")
    dog3 = Dog("Max", 8, "Beagle")
    
    kennel = Kennel()
    kennel.add_dog(dog1)
    kennel.add_dog(dog2)
    kennel.add_dog(dog3)
    
    print("Initial Status:")
    print(kennel)
    for dog in kennel.dogs:
        print(dog)

    oldest = kennel.oldest_dog()
    if oldest:
        print(f"\nOldest dog: {oldest.name} ({oldest.age} years old)")

    labrador_dogs = kennel.find_by_breed("Labrador")
    print("\nLabrador dogs:")
    for dog in labrador_dogs:
        print(dog.name)

    kennel.remove_dog(dog3)
    
    print("\nAfter removing Max:")
    print(kennel)
    for dog in kennel.dogs:
        print(dog)


Initial Status:
Kennel has 3 dogs.
Buddy (5 years old, Golden Retriever)
Bella (3 years old, Labrador)
Max (8 years old, Beagle)

Oldest dog: Max (8 years old)

Labrador dogs:
Bella

After removing Max:
Kennel has 2 dogs.
Buddy (5 years old, Golden Retriever)
Bella (3 years old, Labrador)


# Explaination
    Composition is better here because a Kennel has dogs, but isn't type of dog. It allows more flexibiltiy and easier future changes. Inheritance would be incorrect since a Kennel isn't a "type of" Dog, and would create unnecessary tight coupling between the two.

# Q9

# Task:
    Modules and Packages: Helpers. Create a package helpers/ with modules string_utils.py (function shout(s) uppercases) and math_utils.py (function area(l,w)). Then a main.py that imports with both import ... as ... and from ... import .... Explain namespace collisions and why aliases help. Anchor concepts: modules, packages, import variations, namespaces. Reference: :contentReference[oaicite:15]index=15


# Prompt:
    Write a Python program that creates a package named helpers containing two modules: string_utils.py with a function shout(s) that returns a string in uppercase, and math_utils.py with a function area(l, w) that returns the area of a rectangle. In the main file, import the modules using both import ... as ... and from ... import ..., then demonstrate how aliases help avoid namespace collisions.

In [1]:
#ChatGPT Code

import helpers.string_utils as sutils
import helpers.math_utils as mutils

print(sutils.shout("hello world"))
print(mutils.area(4, 5))

from helpers.string_utils import shout
from helpers.math_utils import area

print(shout("hi there"))
print(area(3, 2))


HELLO WORLD
20
HI THERE
6


# Critique
    Correctness: The code correctly imports the helper modules and calls their functions, producing the right outputs.
    Complexity: O(1) since it only performs simple function calls and arithmetic.
    Robustness: Works for any valid string and numeric inputs. Could add input validation to handle non-string or non-numeric values.
    Readability: Simple and clear. The use of aliases (s, m) makes the code shorter and easier to follow.
    Faithfulness: Fully follows the requirements on modules, packages, and import variations as stated in the task.

In [3]:
#Improved Code

import helpers.string_utils as s
import helpers.math_utils as m

print(s.shout("hello"))
print(m.area(4, 5))

from helpers.string_utils import shout
from helpers.math_utils import area

print(shout("hi"))
print(area(3, 2))


HELLO
20
HI
6


# Explaination
    Namespace collisions happen when two modules have the same function name, which can confuse Python. Aliases fix this by giving each module a short, unique name to keep the code clear and organized.

# Q10

# Task:
    Robust Temperature Converter (Functions + Exceptions). Write two functions: cel_to_far(c) and far_to_cel(f) with type hints and docstrings. Build CLI that asks the user which direction to convert and validates input (gracefully handles bad entries). Include unit tests (doctest or simple asserts) covering typical and edge cases (e.g., -40). Anchor concepts: writing functions, control flow, input handling, testing.

# Prompt
    Write a simple temperature converter with two functions: one to convert Celsius to Fahrenheit and one to convert Fahrenheit to Celsius. Ask the user which conversion they want and handle wrong inputs nicely.

In [24]:
#ChatGPT Code

def cel_to_far(c: float) -> float:
    return (c * 9/5) + 32

def far_to_cel(f: float) -> float:
    return (f - 32) * 5/9

try:
    choice = input("Convert (C)elsius to Fahrenheit or (F)ahrenheit to Celsius? ").strip().lower()
    if choice == "c":
        c = float(input("Enter temperature in Celsius: "))
        print(f"{c}°C = {cel_to_far(c):.2f}°F")
    elif choice == "f":
        f = float(input("Enter temperature in Fahrenheit: "))
        print(f"{f}°F = {far_to_cel(f):.2f}°C")
    else:
        print("Invalid choice.")
except ValueError:
    print("Please enter a number.")

Convert (C)elsius to Fahrenheit or (F)ahrenheit to Celsius?  C
Enter temperature in Celsius:  100


100.0°C = 212.00°F


# Critique
    Correctness: Gives right conversion results.
    Complexity: Very simple (O(1)).
    Robustness: Can handle wrong inputs like letters instead of numbers.
    Readability: Easy to read and understand.
    Faithfulness: Follows the task and shows functions, input handling, and testing.